### PR TITLE
Use `TYPE_CHECKING` for `Study` in `samplers`

### PR DESCRIPTION
**Pull request by nabenabe0928**
## Motivation

As circular references happen in `samplers` when removing `multi_objective` due to `Study`, I replaced `Study` in `samplers`  with `"Study"` using `TYPE_CHECKING`.

## Description of the changes

Replace all `Study` in `samplers` with `"Study"` using `TYPE_CHECKING`.

### CODE REVIEW

**Comment by nabenabe0928** [comment]
Although I made everything `"Study"`, we could also replace it with `optuna.study.Study`.

The former is more concise while the latter allows developers to use the direct link to `Study` in IDE.

I would like to ask which one is better.

**Comment by nabenabe0928** [comment]
@eukaryo @not522 Could you review this PR?

**Comment by not522** [comment]
I recommend `from __future__ import annotations` instead of `"Study"` or `optuna.study.Study`.

File: optuna/study/study.py [hunk 1]
```diff
@@ -36,17 +36,18 @@
...
 _dataframe = _LazyImport("optuna.study._dataframe")
 
 if TYPE_CHECKING:
     from optuna.study._dataframe import pd
+    from optuna.trial import FrozenTrial
+    from optuna.trial import Trial
 
 
-ObjectiveFuncType = Callable[[trial_module.Trial], Union[float, Sequence[float]]]
+ObjectiveFuncType = Callable[["Trial"], Union[float, Sequence[float]]]
```
Comment (not522) [inline optuna/study/study.py:50]: How about moving it under `if TYPE_CHECKING:`?

```suggestion
    ObjectiveFuncType = Callable[[Trial], Union[float, Sequence[float]]]
```

Comment (nabenabe0928) [reply optuna/study/study.py]: Oh, I did not know that it works in this way!
I reflected your suggestion!

**Review by eukaryo — COMMENTED** [review]
LGTM.

**Review by eukaryo — APPROVED** [review]

**Pull request merged**